### PR TITLE
feat: allow setting channel and credential providers in DatastoreOptions

### PR DESCRIPTION
**Pull request by kolea2**
Similar to Firestore, https://github.com/googleapis/java-firestore/blob/main/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreOptions.java#L162-L168, this allows users to configure more custom options available in gax and gRPC like credentials providers and channel pooling.

### CODE REVIEW

**Review by bhshkh — APPROVED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java [hunk 1]
```diff
@@ -84,20 +94,50 @@ public static class Builder extends ServiceOptions.Builder<Datastore, DatastoreO
...
+    private TransportChannelProvider channelProvider = null;
+    private CredentialsProvider credentialsProvider = null;
 
     private Builder() {}
 
     private Builder(DatastoreOptions options) {
       super(options);
       namespace = options.namespace;
       databaseId = options.databaseId;
+      this.channelProvider = options.channelProvider;
```
Comment (bhshkh) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:106]: Nit: the above 2 variables don't have "this." but the new ones do.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java [hunk 2]
```diff
@@ -119,6 +159,32 @@ private DatastoreOptions(Builder builder) {
...
     namespace = MoreObjects.firstNonNull(builder.namespace, defaultNamespace());
     databaseId = MoreObjects.firstNonNull(builder.databaseId, DEFAULT_DATABASE_ID);
+
+    // todo see if we can update this
+    if (getTransportOptions() instanceof HttpTransportOptions
+        && (builder.channelProvider != null || builder.credentialsProvider != null)) {
+      throw new IllegalArgumentException(
+          "Only gRPC transport allows setting of channel provider or credentials provider");
+    } else {
+      this.channelProvider =
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:169]: Do we still want to instantiate this to be GrpcTransportProvider if `(getTransportOptions() instanceof HttpTransportOptions) && (builder.channelProvider == null && builder.credentialsProvider == null) == true`?

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreOptionsTest.java [hunk 3]
```diff
@@ -48,6 +64,16 @@ public void setUp() {
             .setServiceRpcFactory(datastoreRpcFactory)
             .setProjectId(PROJECT_ID)
             .setDatabaseId(DATABASE_ID)
+            .setCredentials(NoCredentials.getInstance())
+            .setHost("http://localhost:" + PORT);
+
+    options2 =
```
Comment (cindy-peng) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreOptionsTest.java:70]: Is `options2` only for testCustomChannelAndCredentials()? Do we need both `options` and `options2` to cover all the test cases? If not, could we rename them to be more meaningful?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java (hunk 1 above)
Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:106]: thanks, fixed to have the above two to also have `this.` prefix

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java (hunk 2 above)
Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:169]: great catch, thank you! I updated the logic and added a test to confirm it's working as expected

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreOptionsTest.java (hunk 3 above)
Comment (kolea2) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreOptionsTest.java:70]: inlined to the specific test

**Review by cindy-peng — APPROVED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java [hunk 4]
```diff
@@ -69,6 +75,10 @@ public ServiceRpc create(DatastoreOptions options) {
         if (options.getTransportOptions() instanceof GrpcTransportOptions) {
           return new GrpcDatastoreRpc(options);
         } else if (options.getTransportOptions() instanceof HttpTransportOptions) {
+          // todo see if we can remove this check
+          if (DatastoreUtils.isEmulator(options)) {
+            throw new IllegalArgumentException("Only GRPC channels are allowed for emulator.");
```
Comment (meltsufin) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:80]: We use the emulator in Spring Cloud GCP with REST. Maybe [this](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/11c070e2ba0bc17b5ba628ed763bb8d0144ade5a/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java#L93) could be helpful?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java [hunk 5]
```diff
@@ -84,20 +94,50 @@ public static class Builder extends ServiceOptions.Builder<Datastore, DatastoreO
...
 
     private Builder() {}
 
     private Builder(DatastoreOptions options) {
       super(options);
-      namespace = options.namespace;
-      databaseId = options.databaseId;
+      this.namespace = options.namespace;
+      this.databaseId = options.databaseId;
+      this.channelProvider = options.channelProvider;
```
Comment (meltsufin) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:106]: Why aren't you using the `setChannelProvider()` here? It has additional checks.

Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java:106]: `setChannelProvider` returns a Builder, but I can extract out the validation and use it in both places.

Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java]: added!

**Review by meltsufin — APPROVED** [review]

**Pull request merged**